pkg/operator/controller/ingress/load_balancer_service.go

```diff
@@ -751,17 +776,88 @@ func loadBalancerServiceIsProgressing(ic *operatorv1.IngressController, service
+			if clbParams := getAWSClassicLoadBalancerParametersInSpec(ic); clbParams != nil {
+				wantSubnets = clbParams.Subnets
+			}
+			if clbParams := getAWSClassicLoadBalancerParametersInStatus(ic); clbParams != nil {
+				haveSubnets = clbParams.Subnets
+			}
+			paramsFieldName = "classicLoadBalancer"
+		}
+		if !compareAWSSubnets(wantSubnets, haveSubnets) {
+			// Generate JSON for the oc patch command as well as "pretty" json
+			// that will be used for a more human-readable error message.
+			haveSubnetsPatchJson := convertAWSSubnetListToPatchJson(haveSubnets, "null", "null")
+			haveSubnetsPrettyJson := convertAWSSubnetListToPatchJson(haveSubnets, "{}", "[]")
+			wantSubnetsPrettyJson := convertAWSSubnetListToPatchJson(wantSubnets, "{}", "[]")
+			err := fmt.Errorf("The IngressController subnets were changed from %q to %q.", haveSubnetsPrettyJson, wantSubnetsPrettyJson)
+			ocPatchRevertCmd := fmt.Sprintf("oc -n openshift-ingress-operator patch ingresscontrollers/%[1]s --type=merge --patch='{\"spec\":{\"endpointPublishingStrategy\":{\"type\":\"LoadBalancerService\",\"loadBalancer\":{\"providerParameters\":{\"type\":\"AWS\",\"aws\":{\"type\":\"%[2]s\",\"%[3]s\":{\"subnets\":%[4]s}}}}}}}'", ic.Name, getAWSLoadBalancerTypeInStatus(ic), paramsFieldName, haveSubnetsPatchJson)
+			err = fmt.Errorf("%[1]s  To effectuate this change, you must delete the service: `oc -n %[2]s delete svc/%[3]s`; the service load-balancer will then be deprovisioned and a new one created. This will most likely cause the new load-balancer to have a different host name and IP address and cause disruption. To return to the previous state, you can revert the change to the IngressController: `%[4]s`", err.Error(), service.Namespace, service.Name, ocPatchRevertCmd)
+			errs = append(errs, err)
+		}
+	}
+
 	errs = append(errs, loadBalancerSourceRangesAnnotationSet(service))
 	errs = append(errs, loadBalancerSourceRangesMatch(ic, service))
 
 	return kerrors.NewAggregate(errs)
 }
 
+// convertAWSSubnetListToPatchJson converts an AWSSubnets object to a JSON formatted string
+// to build an oc patch command. It defaults nil or no subnet values to emptySubnetValue
+// and empty ids or names slices to emptySubnetSliceValue.
+func convertAWSSubnetListToPatchJson(subnets *operatorv1.AWSSubnets, emptySubnetValue, emptySubnetSliceValue string) string {
+	// If subnets are nil, or an empty list, return the emptySubnetValue.
+	if subnets == nil || (len(subnets.Names) == 0 && len(subnets.IDs) == 0) {
+		return emptySubnetValue
+	}
+
+	// Default subnet names and IDs to emptySubnetSliceValue.
+	namesJSON, idsJSON := emptySubnetSliceValue, emptySubnetSliceValue
+
+	// Marshal names and IDs separately.
+	if len(subnets.Names) > 0 {
+		var err error
+		namesJSONBytes, err := json.Marshal(subnets.Names)
+		if err != nil {
+			log.Error(err, "error marshaling subnet names")
+			return ""
+		}
+		namesJSON = string(namesJSONBytes)
+	}
+	if len(subnets.IDs) > 0 {
+		var err error
```

Comment:
why do need to declare var err because we already have initialised it down err := ?